File: Models/Product.cs
using Hazelcast.Serialization;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace HazelcastCaching.Models
{
    public class Product : IIdentifiedDataSerializable
    {
        public const int ClassId = 100;

        [BsonId]
        public ObjectId Id { get; set; }
        public string Code { get; set; }
        public string Name { get; set; }
        public float Price { get; set; }
        public string Description { get; set; }

        public void ReadData(IObjectDataInput input)
        {
            Name = input.ReadString();
            Price = input.ReadFloat();
            Description = input.ReadString();
        }

        public void WriteData(IObjectDataOutput output)
        {
            output.WriteString(Name);
            output.WriteFloat(Price);
            output.WriteString(Description);
        }

        public int FactoryId => ProductSerializableFactory.FactoryId;
        int IIdentifiedDataSerializable.ClassId => 100;
    }
}

File: Java classes/Product.java
import java.io.IOException;

import com.hazelcast.nio.ObjectDataInput;
import com.hazelcast.nio.ObjectDataOutput;
import com.hazelcast.nio.serialization.IdentifiedDataSerializable;

public class Product implements IdentifiedDataSerializable {
	
	static final int CLASS_ID = 100;
	
	public String name;
	public float price;
	public String description;
	
	public Product()
	{
	}
	
	public Product (String name, float price, String description)
	{
		this.name=name;
		this.price=price;
		this.description=description;
	}

	@Override
	public void readData(ObjectDataInput in) throws IOException {
		name = in.readString();
		price = in.readFloat();
		description = in.readString();
		
	}

	@Override
	public void writeData(ObjectDataOutput out) throws IOException {
		out.writeString(name);
		out.writeFloat(price);
		out.writeString(description);
		
	}

	@Override
	public int getClassId() {
		return CLASS_ID;
	}

	@Override
	public int getFactoryId() {
		return ProductSerializableFactory.FACTORY_ID;
	}
}
